### DISCUSSION TITLE
Src block hiliting via polymode

### DISCUSSION THREAD

**Discussion by TobiasZawada** [emacsorphanage/adoc-mode#20] [Show and tell]
Hi @bbatsov ,
Branch [`SrcBlock/polymode`](https://github.com/bbatsov/adoc-mode/tree/SrcBlock/polymode) demonstrates the usage of `polymode` for source block highlighting.

A minimal example is `README.adoc`. Source blocks for `emacs-lisp` are correctly handled with `emacs-lisp-mode`. But, for source blocks with language `shellsession` the fallback `prog-mode` is used.

This  branch is just a concept study.
The other two concepts are:

- use `mmm-mode`
- adapt Org source blocks to `adoc-mode`

**Comment by bbatsov** [comment]
What's the approach that `markdown-mode` is taking? I was thinking we should probably emulate whatever they are doing.

**Reply by TobiasZawada** [reply]
[`markdown-mode`](https://github.com/jrblevin/markdown-mode) uses the third approach, i.e., adopt the [Org mode approach](https://github.com/jrblevin/markdown-mode/blob/92d037504742cb3e579f1f488bc3fc23c648ff29/markdown-mode.el#L8787-L8822).

**Reply by bbatsov** [reply]
Got it. Well, if it's not more complex my preference would be for us to do the same.

**Comment by TobiasZawada** [comment]
I would love if we could convince the people from Org and Md to use a more general code for source blocks in general markdown modes. (I hate copying code.)
Also the source block execution and the editing in dedicated buffers could be generalized for all general markdown modes.

I know the attitude of the Org people to reject such proposals with performance arguments as excuse.

But, a library of that kind could exploit macros and aliases that are directly compiled into the target code.
So, I do not really agree with such arguments.

**Reply by bbatsov** [reply]
Well, I get your point, and I'm open trying trying `polymode`, but given its size we'd probably have to make it an optional dependency. 

P.S. Copying code is not always as bad as it sounds - some times it's better to have no deps and copying code is how you get there. :-) Every dependency is a long-term liability, so I reach for those sparingly.

**Reply by TobiasZawada** [reply]
Branch `SrcBlock/Org` is now a first try for the Org approach.
Some known issues:
- fontification fails when creating a code block from scratch, but update with `font-lock-fontify-buffer` works
- attribute line and delimiters are hilited by `markup-meta-face` and not by `markup-block-del`
- editing of the source code works but after editing the source language or the delimiters highlighting is gone
- `adoc-code-block-begin-regexp` should probably be done with proper `rx`

**Reply by TobiasZawada** [reply]
@bbatsov After the last commit  edf1bd4b3792d00a3bee753ed27aefffe809f326 source block highlighting should become quite usable. Could you test it? (I already did some rudimentary tests.)

Interesting question is: How bad is the limitation of the search distance for code block headers? I think such limitation is needed for performance reasons, but it comes with the caveat of false fontification of long source blocks after edits.

**Reply by TobiasZawada** [reply]
@bbatsov Next logical thing is native indentation.

**Reply by bbatsov** [reply]
Can you open some PR with your progress so far? If it's more visible we might find more testers and get more feedback as well.

**Reply by TobiasZawada** [reply]
> Can you open some PR with your progress so far? If it's more visible we might find more testers and get more feedback as well.

Fine by me.  I assume that you review the changes in the pull-request.

**Reply by TobiasZawada** [reply]
@bbatsov Thank you very much for the review. I've corrected all except the `markup-faces.el`-thing.

**Reply by TobiasZawada** [reply]
Note, that the method from Org-mode makes editing of large source blocks slow.
One could keep the temporary buffers of code blocks whose size exceed a certain limit alive and apply only the changes from the adoc-mode buffer to those temp buffer. Then only a part of the code block needs to be re-fontified.

**Reply by bbatsov** [reply]
> Note, that the method from Org-mode makes editing of large source blocks slow.

I think it's uncommon to inline large code blocks, so I guess this won't a problem most of the time. Still, we can mention it in the docs for this feature.
